<a href="https://colab.research.google.com/github/Dimas0824/Machine_Learning/blob/main/Jobsheet_13/ML_WEEK13.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Praktikum 1

In [1]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 2
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.25491446917804494
Epoch 1000, Loss: 0.20602602264180095
Epoch 2000, Loss: 0.16742418599311107
Epoch 3000, Loss: 0.09097069374601213
Epoch 4000, Loss: 0.025433652379636476
Epoch 5000, Loss: 0.0115355492288669
Epoch 6000, Loss: 0.007023129122780075
Epoch 7000, Loss: 0.0049360862072992395
Epoch 8000, Loss: 0.003763875630922241
Epoch 9000, Loss: 0.0030227066091228646
Prediksi:
[[0.04274377]
 [0.95222173]
 [0.95245898]
 [0.06077781]]


## Tugas 1:

* Ubah jumlah neuron hidden layer menjadi 3.

* Bandingkan hasil loss dengan konfigurasi awal.

In [2]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 3
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.267429293447181
Epoch 1000, Loss: 0.2335236255859649
Epoch 2000, Loss: 0.07240969364074616
Epoch 3000, Loss: 0.017021822040387657
Epoch 4000, Loss: 0.008065094005131578
Epoch 5000, Loss: 0.005035932533295781
Epoch 6000, Loss: 0.0035885835046952815
Epoch 7000, Loss: 0.002758691713314847
Epoch 8000, Loss: 0.002226763684095194
Epoch 9000, Loss: 0.001859330592831659
Prediksi:
[[0.04428184]
 [0.96339008]
 [0.95782466]
 [0.03587744]]


dari hasil diatas, model dengan menggunakan 2 hidden layer lebih stabil diawal tapi konvergensi lebih lambat daripada model 3 hidden layer. Sedangkan model dengan 3 hidden layer memiliki loss awal lebih tinggi daripada model 2 hidden layer, tapi setelah 2000 epoch performa model jauh lebih baik dan konvergen lebih cepat.

* Tambahkan fungsi aktivasi ReLU dan bandingkan hasil.

In [3]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 3
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

def relu(x):
    return np.maximum(0, x)

def relu_derivative(x):
    return (x > 0).astype(float)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = relu(z1)   # gunakan ReLU di hidden layer
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)  # output tetap sigmoid

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * relu_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss:.6f}")

# Output akhir lebih rapi
print("\nPrediksi akhir:")
for i in range(len(X)):
    print(f"Input: {X[i]} -> Target: {y[i][0]} | Prediksi: {a2[i][0]:.4f}")

Epoch 0, Loss: 0.223214
Epoch 1000, Loss: 0.126319
Epoch 2000, Loss: 0.125457
Epoch 3000, Loss: 0.125249
Epoch 4000, Loss: 0.125171
Epoch 5000, Loss: 0.125128
Epoch 6000, Loss: 0.125107
Epoch 7000, Loss: 0.125084
Epoch 8000, Loss: 0.125072
Epoch 9000, Loss: 0.125062

Prediksi akhir:
Input: [0 0] -> Target: 0 | Prediksi: 0.4999
Input: [0 1] -> Target: 1 | Prediksi: 0.9906
Input: [1 0] -> Target: 1 | Prediksi: 0.4999
Input: [1 1] -> Target: 0 | Prediksi: 0.0121


Model dengan tambahan activation ReLU, memiliki loss tidak turun mendekati nol, melainkan stagnan dan pletau di sekitar 0.125. Ini menunjukkan model tidak dapat belajar pola data XOR. Pada model tanpa ReLU, prediksi mendektai target dengan baik. Sedangkan dengan ReLU activation model juga hampir benar tapi memiliki kesalahan yang lebih tinggi daripada model sebelumnya.